dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,42 +52,208 @@ public Pattern getParameterPattern() {
-                        columnName,
-                        Direction.INPUT,
-                        typeMapper.getJDBCSQLTypeForDBType(dataType),
-                        getJavaClass(dataType), position++);
-                allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+            try (ResultSet rs = getColumns()) {
+                while (rs.next()) {
+                    String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+                    allParams.put(NameNormaliser.normaliseName(paramName),
+                                  createColumnAccessor(paramName, rs));
+                }
             }
-            rs.close();
             return allParams;
         }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, columnDirection(rs));
+        }
+
+        protected Direction columnDirection(ResultSet rs) throws SQLException {
+            Direction columnDir = getColumnDirection(rs);
+            if (columnDir == null) {
+                throw new SQLException("Invalid column/parameter type");
+            }
+            return columnDir;
+        }
+
+        abstract protected Direction getColumnDirection(ResultSet rs) throws SQLException;
+
+        protected DbParameterAccessor createDbParamAccessor(String paramName, ResultSet rs, Direction direction)
+                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            return createDbParameterAccessor(paramName, direction,
+                           typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                           getJavaClass(paramTypeName), rs.getInt("ORDINAL_POSITION") - 1);
+        }
+    }
+
+    private class DatabaseTable extends DatabaseObject {
+
+        private DatabaseTable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getColumnDirection(ResultSet rs) {
+            return Direction.INPUT;
+        }
+    }
+
+    abstract private class DatabaseExecutable extends DatabaseObject {
+
+        private DatabaseExecutable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected Direction getColumnDirection(ResultSet rs) throws SQLException {
+            return columnTypeToDirection(rs.getInt("COLUMN_TYPE"));
+        }
+
+        abstract protected Direction columnTypeToDirection(int columnType);
```

Comment:
> Maybe at the cost of little duplication we could take a step back and move that down to the children: `switch(rs.getInt("COLUMN_TYPE"))`. Or alternatively declare a `getColumnTypeId(ResultSet)` inside `DatabaseObject` (despite that we don't use it for tables). Maybe it's a good tradeoff since this way we can keep the class hierarchy 2-levels only and perhaps drop the `int->Direction` mapping function.

I'm not sure about it. I think my own preferred version was actually where we only had any methods relating to `Direction` in `DatabaseExecutable` and below.